parity/main.rs

```diff
@@ -18,53 +18,86 @@
 
 #![warn(missing_docs)]
 
-extern crate parity;
-
 extern crate ctrlc;
 extern crate dir;
 extern crate fdlimit;
 #[macro_use]
 extern crate log;
 extern crate panic_hook;
+extern crate parity;
 extern crate parking_lot;
 
 #[cfg(windows)] extern crate winapi;
 
-use std::{process, env};
-use std::io::{self as stdio, Read, Write};
-use std::fs::{remove_file, metadata, File, create_dir_all};
-use std::path::PathBuf;
-use std::sync::Arc;
 use ctrlc::CtrlC;
 use dir::default_hypervisor_path;
 use fdlimit::raise_fd_limit;
 use parity::{start, ExecutionAction};
 use parking_lot::{Condvar, Mutex};
+use std::fs::{remove_file, metadata, File, create_dir_all};
+use std::io::{self as stdio, Read, Write};
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::{process, env, ffi::OsString};
+
+const PLEASE_RESTART_EXIT_CODE: i32 = 69;
+
+#[derive(Debug)]
+enum Error {
+	BinaryNotFound,
+	StatusCode(i32),
+	Restart,
+	UnknownStatusCode,
+}
 
-fn updates_path(name: &str) -> PathBuf {
-	let mut dest = PathBuf::from(default_hypervisor_path());
+fn update_path(name: &str) -> PathBuf {
+	let mut dest = default_hypervisor_path();
 	dest.push(name);
 	dest
 }
 
-fn latest_exe_path() -> Option<PathBuf> {
-	File::open(updates_path("latest")).ok()
-		.and_then(|mut f| { let mut exe = String::new(); f.read_to_string(&mut exe).ok().map(|_| updates_path(&exe)) })
+fn latest_exe_path() -> Result<PathBuf, Error> {
+	File::open(update_path("latest")).and_then(|mut f| { 
+			let mut exe = String::new(); 
+			trace!(target: "updater", "latest binary path: {:?}", f); 
+			f.read_to_string(&mut exe).map(|_| update_path(&exe))
```

Comment:
There are two files in `.local/share/io.parity.ethereum-updates` (path for *NIX-based systems):
- latest - contains the name of the latest binary 
- the actual binary - some like `parity-1.9.5-ff821daf1da42865f229aee35f2e74e7b2dd8db2`

So if `.local/share/io.parity.ethereum-updates/latest` exists it will return the path the `newest binary` e.g, `.local/share/io.parity.ethereum-updates/parity-1.9.5-ff821daf1da42865f229aee35f2e74e7b2dd8db2`

Reply:
Changing the name of the variable from `exe` to `exe_path` would make it simple to understand :)